Add support for jitDispatchJ9Method for MH on P

matthewhall2 opened the pull request:
Enables transformation of invokeBasic, linkToStatic, and linkToSpecial calls into jitDispatchJ9Method acall nodes. Codegen recognizes this node in private linkage and produces code to directly call the MethodHandletarget at runtime
- using a new PPCArgFlushHelperCallSnippet snippet when the
              target is interpreted
 - otherwise branch and link to jit start pc of target

matthewhall2 commented:
@zl-wang can I get a review on this please? This is what we discussed with @IBMJimmyk last week

zl-wang commented:
High-level comments:
1) HelperCallSnippet seemed being used for disjoint purposes across platforms and in OMR side as well. this use (new file) in OpenJ9 PPC codegen looked more aligned with OMR side of story. Please consider how to make it more readable/understandable and less confusing for future newcomers!  I was thinking:  could we do away without this new file and integrating MH invoke support in existing files (anyway)?  or, naming the new file for its specific purpose (MH invocation support)?
2) rightToLeft or leftToRight argument ordering should be still applicable to the invocation by and large, instead of disabling it as you wrote it, although i understood this thing has never been used in reality to begin with.  You only need to special-case the first argument (j9method? if I remembered correctly). The shifted argument list should be identical to a normal java method dispatch in IL representation ... so, the disabling is unexpected at high level.

I will review the details when i get around.

matthewhall2 committed:
Add support for jitDispatchJ9Method for MH on P

    Enables transformation of invokeBasic, linkToStatic, and linkToSpecial
    calls into jitDispatchJ9Method acall nodes. Codegen recognizes this node
    in private linkage and produces code to directly call the MethodHandle
    target at runtime
        - using a new PPCArgFlushHelperCallSnippet snippet when the
          target is interpreted
        - otherwise branch and link to jit start pc of target

Signed-off-by: Matthew Hall <matthew.hall3@outlook.com>

zl-wang commented on runtime/compiler/p/codegen/J9PPCSnippet.hpp:
```diff
@@ -68,6 +68,23 @@ class PPCReadMonitorSnippet : public TR::PPCHelperCallSnippet {
     int32_t getLoadOffset() { return _loadOffset; }
 };
 
```
PPCJ9HelperCallSnippet sounds like this is good for any generic J9 helper call, but, in reality, this is heavily tailored for jitDispatchJ9Method special case. I would suggest having a name closely corresponding to its true purpose and suitability.

zl-wang reviewed (commented):
will continue once i get some cycles.

zl-wang commented on runtime/compiler/p/codegen/PPCPrivateLinkage.cpp:
```diff
@@ -1432,6 +1434,11 @@ int32_t J9::Power::PrivateLinkage::buildPrivateLinkageArgs(TR::Node *callNode,
 
+    if (isJitDispatchJ9Method) {
+        specialArgReg = getProperties().getJ9MethodArgumentRegister();
+        TR::addDependency(dependencies, NULL, TR::RealRegister::gr0, TR_GPR, cg());
```
this addDependency for gr0 appeared very abrupt. I don't see the necessity to confuse the readers. you made it like gr0 being some special register to jitDispatchJ9Method exclusively, but it is not. Later on, all volatile registers are added to the dependency anyway.

zl-wang commented on runtime/compiler/p/codegen/PPCPrivateLinkage.cpp:
```diff
@@ -1447,6 +1454,7 @@ int32_t J9::Power::PrivateLinkage::buildPrivateLinkageArgs(TR::Node *callNode,
         totalSize += TR::Compiler->om.sizeofReferenceAddress();
     }
 
+    // will not process special args
```
name the variable with a more meaningful one and comment the above with a fuller description. capture these below:
1) if there is a special child, it is always the first. it will be processed outside the loop below, as `from ` value was skipped already above;
2) special child is not an argument to the J9Method, but might be used (in different ways under different situations) in calling the J9Method concerned;

zl-wang commented on runtime/compiler/p/codegen/PPCPrivateLinkage.cpp:
```diff
@@ -1526,7 +1535,10 @@ int32_t J9::Power::PrivateLinkage::buildPrivateLinkageArgs(TR::Node *callNode,
             case TR::Int32:
             case TR::Address: // have to do something for GC maps here
-                if (i == firstArgumentChild && callNode->getOpCode().isIndirect()) {
+                              // first child holds the J9MethodPointer for jitDispatchJ9Method, want to pushAddressArg
```
please align the comment

zl-wang commented on runtime/compiler/p/codegen/PPCPrivateLinkage.cpp:
```diff
@@ -1771,8 +1785,11 @@ int32_t J9::Power::PrivateLinkage::buildPrivateLinkageArgs(TR::Node *callNode,
-    if (!dependencies->searchPreConditionRegister(TR::RealRegister::gr11))
+    if (!dependencies->searchPreConditionRegister(TR::RealRegister::gr11)) {
+        if (isJitDispatchJ9Method)
+            TR_ASSERT_FATAL(false, "r11 dependency should have been added\n");
```
i don't see the necessity to modify the linkage properties, since you are explicitly associating gr11 with J9Method here anyway.  it looks to me simpler that you use gr11 directly without going through the properties symbolic retriever and the above-below code can be simplified further ... at least easier to read.

zl-wang commented on runtime/compiler/p/codegen/PPCPrivateLinkage.cpp:
```diff
@@ -2736,19 +2753,86 @@ void J9::Power::PrivateLinkage::buildDirectCall(TR::Node *callNode, TR::SymbolRe
+    if (!isJitDispatchJ9Method
+        && (callSymbol->isJITInternalNative()
             || (!callSymRef->isUnresolved() && !callSymbol->isInterpreted()
                 && ((forceUnresolvedDispatch && callSymbol->isHelper()) || !forceUnresolvedDispatch)))) {
```
recommend tidying up the existing clumsy code (not due to your code though):
`((forceUnresolvedDispatch && callSymbol->isHelper()) || !forceUnresolvedDispatch)`   is equivalent to
`(callSymbol->isHelper() || !forceUnresolvedDispatch)`

zl-wang commented on runtime/compiler/p/codegen/J9PPCSnippet.cpp:
```diff
@@ -292,6 +293,45 @@ int32_t TR::PPCReadMonitorSnippet::setEstimatedCodeLocation(int32_t estimatedSni
+        // move value in r11 to r3 for the interpreter
+        // or     r11   r3    r11    444        0
+        // 011111 01011 00011 01011  0110111100 0
+        *(int32_t *)buffer = 0x7D635B78;
```
strongly recommend doing this symbolically for better readability ... referring to the previous code in this file. Dragging the binary encoding and calling the corresondingly named routines to fill in RA/RS/RB

zl-wang commented on runtime/compiler/p/codegen/PPCPrivateLinkage.cpp:
```diff
@@ -2736,19 +2753,86 @@ void J9::Power::PrivateLinkage::buildDirectCall(TR::Node *callNode, TR::SymbolRe
+
+        TR::LabelSymbol *startICFLabel = generateLabelSymbol(cg());
+        TR::LabelSymbol *doneLabel = generateLabelSymbol(cg());
+        TR::LabelSymbol *oolLabel = generateLabelSymbol(cg());
```
this label is not used anymore. better to delete it?

zl-wang commented on runtime/compiler/p/codegen/PPCPrivateLinkage.cpp:
```diff
@@ -2736,19 +2753,86 @@ void J9::Power::PrivateLinkage::buildDirectCall(TR::Node *callNode, TR::SymbolRe
+        // slowCallOOL->swapInstructionListsWithCompilation();
+
+        generateLabelInstruction(cg(), TR::InstOpCode::label, callNode, startICFLabel);
+
```
no consideration for 32bit JVM anymore? i guessed jitDispatchJ9Method only applicable to post-Java8?

zl-wang commented on runtime/compiler/p/codegen/PPCPrivateLinkage.cpp:
```diff
@@ -2736,19 +2753,86 @@ void J9::Power::PrivateLinkage::buildDirectCall(TR::Node *callNode, TR::SymbolRe
+        // compiled - jump to jit entry point
+        generateTrg1MemInstruction(cg(), TR::InstOpCode::lwz, callNode, j9MethodReg,
+            TR::MemoryReference::createWithDisplacement(cg(), scratchReg, -4,
+                TR::Compiler->om.sizeofReferenceAddress()));
```
although it doesn't matter (from memory aliasing perspective), here it should be hard-coded as 4 (since lwz really accesses 4-byte in memory).

zl-wang commented on runtime/compiler/p/codegen/PPCPrivateLinkage.cpp:
```diff
@@ -2736,19 +2753,86 @@ void J9::Power::PrivateLinkage::buildDirectCall(TR::Node *callNode, TR::SymbolRe
+            TR::MemoryReference::createWithDisplacement(cg(), scratchReg, -4,
+                TR::Compiler->om.sizeofReferenceAddress()));
+        generateShiftRightLogicalImmediate(cg(), callNode, j9MethodReg, j9MethodReg, 16);
+        generateSignExtendInstruction(callNode, j9MethodReg, j9MethodReg, cg(), 4);
```
this sign-extension is a total waste, since there are at most 21 arguments passed in registers, i.e. the offset is <= 84 bytes. completely ok to delete it. by the way, for correctness, it should be two-byte zero extension too (since the value is 16bits to begin with and always zero or positive.

zl-wang commented on runtime/compiler/p/codegen/PPCPrivateLinkage.cpp:
```diff
@@ -2736,19 +2753,86 @@ void J9::Power::PrivateLinkage::buildDirectCall(TR::Node *callNode, TR::SymbolRe
+        generateSrc1Instruction(cg(), TR::InstOpCode::mtctr, callNode, scratchReg2);
+        gcPoint = generateInstruction(cg(), TR::InstOpCode::bctrl, callNode);
+        gcPoint->PPCNeedsGCMap(regMapMask);
+        cg()->stopUsingRegister(scratchReg2);
```
i don't follow your logic here to stopUsing scratchReg2 alone.  all of the registers on the dependencies would be taken care of their live-ness later on or somewhere i supposed. no need to specially treat scratchReg2 here.

zl-wang commented on runtime/compiler/p/codegen/J9PPCSnippet.cpp:
```diff
@@ -292,6 +293,49 @@ int32_t TR::PPCReadMonitorSnippet::setEstimatedCodeLocation(int32_t estimatedSni
+        TR::RealRegister *interpreterReg = cg()->machine()->getRealRegister(TR::RealRegister::gr3);
+
+        TR::InstOpCode opcode;
+        opcode.setOpCodeValue(TR::InstOpCode::or_r);
```
`TR::InstOpCode::OR `is better than `TR::InstOpCode::or_r` (**or instruction record form**) here, since you are not expecting `CR0` to be set and you have no use of that CR0 result anyway.

zl-wang reviewed (changes requested):
otherwise, it looks good to me. once you changed that, i can proceed to approve.

zl-wang reviewed (approved):
LGTM

zl-wang commented:
Jenkins test sanity aix,plinux jdk8,jdk21,jdk25

zl-wang commented:
JDK8 tests all had failure (similar symptom: crashing in interpreter on both aix and ppc64le). JDK25 on ppc64le had a single JFR failure ... crashing in iterating virtual-thread continuation objects and relation to memory management. 

I suspected none of these has any relation to this PR, but @matthewhall2 might want to check them out in detail to make sure.

zl-wang commented:
i figured this IL transformation (i,e creating jitDispatchJ9Method call node) is still to be enabled on Power.  i.e. as it is, there is no such IL at all (on p, that is, in all those tests above). catching any regression was the purpose, of course.

matthewhall2 commented:
yes, it is enbabled in this PR, but not on java 8. changing the register used in PPCPrivateLinkage.cpp:1425 fixes the segfault (the old methodhandle implementation uses that), but introduces an illegal instruction error. Testing some fixes now to see why these affect java 8 (in another branch to keep noise out the the PR) some of the  changes in buildargs may have affected it

zl-wang commented:
that register is good for post-JDK8, but not for JDK8 itself (plus, it is not expected to be active in JDK8 at all).  that sounds ominous to me.  the crashing reason identification is likely not in the right direction.

for dispatchVirtual (to interpreter), certain VM helper (you might need to talk to GAC about the detail) actually expected the VFT offset in gr12 (on Power).  or, some other VM helper looked back into the JIT-ed code to find the VFT offset (and there is none in this case).  you probably need to investigate to make sure none of these will happen in these cases.

zl-wang commented:
since the particular J9Method is known in your cases for each running instance, I am pretty confident you can make it as the case where you don't need any of the generic situations mentioned above.
